components/common/src/command/package/install.rs

```diff
@@ -843,22 +839,33 @@ impl<'a> InstallTask<'a> {
                                -> Result<()>
         where T: UIWriter
     {
-        ui.status(Status::Downloading, ident)?;
-        match self.api_client
-                  .fetch_package((ident.as_ref(), target),
-                                 token,
-                                 self.artifact_cache_path,
-                                 ui.progress())
-                  .await
+        if let Err(e) = retry_builder_api!(None, async {
+                            self.api_client
+                                .fetch_package((ident.as_ref(), target),
+                                               token,
+                                               self.artifact_cache_path,
+                                               ui.progress())
+                                .await
+                        }).await
         {
-            Ok(_) => Ok(()),
-            Err(api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
-                println!("Host platform or architecture not supported by the targeted depot; \
-                          skipping.");
-                Ok(())
-            }
-            Err(e) => Err(Error::from(e)),
+            return Err(Error::DownloadFailed(format!("When suitable, we try \
+                                                      once then re-attempt {} \
+                                                      times with a back-off \
+                                                      algorithm. Unfortunately, \
+                                                      it seems we still could \
+                                                      not download {} for {}. \
+                                                      (Some HTTP error \
+                                                      conditions are not \
+                                                      practically worth \
+                                                      retrying) - last error: \
+                                                      {}.",
+                                                     API_RETRIES,
+                                                     ident,
+                                                     target,
+                                                     e)));
```

Comment:
These `DownloadFailed` messages are very similar. Instead of having `DownloadFailed` take a `String` what if it took an `enum` describing what kind of download was attempted (eg `Package`, `OriginKey`, `EncryptionKey ...). Each of those enums could contain the relevant information for the failure (eg `Package` would have the `PackageIdent`). Then these long strings would mostly live in the `fmt::Display` `impl` of `Error`.

Reply:
This is a superb idea. Added in https://github.com/habitat-sh/habitat/pull/7874/commits/2a457e8b6f09ace44d7248a5f5146dcfbd2bb44e